### PR TITLE
Exceptions: do not rethrow from functions with incompatible exception specifications.

### LINKED ISSUE #159: `A15-4-2`: Only report the first `noexcept` function that throws
### Affected rules
 - `A15-4-2`
 - `ERR55-CPP`

### Description

`noexcept` functions frequently call other `noexcept` functions. If a `noexcept` function throws an exception, then we currently report that as a violation in every `noexcept` function that calls that function. This creates multiple alerts for one issues. We should instead report the original instance and not the others.

### Example
```cpp
void f1() noexcept {
  throw foo(); // report this case
}

void f2() noexcept {
  f1(); // do not report this case
}
```

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Per `[except.spec]`, by default if an exception specification is violated, the exception is not rethrown and `std::terminate` is called. I have therefore modified the `ExceptionFlow.qll` library to no longer propagate exceptions thrown by `noexcept` or dynamic specification functions where the exception type isn't listed. This should significant reduce the number of false positives reported by this rule.

Note: this has impacted a number of different queries that used the base exception flow library.

Fixes #159.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
   - `A15-4-2`, `ERR55-CPP` - reduce false positives for `noexcept` functions which call other `noexcept` function which may throw.
   - `A15-2-2` - reduce false positives for constructors which call `noexcept` functions.
   - `A15-4-5` - reduce false positives for checked exceptions that are thrown from `noexcept` functions called by the original function.
   - `DCL57-CPP` - do not report exceptions thrown from `noexcept` functions called by deallocation functions or destructors.
   - `A15-5-1`, `M15-3-1` - do not report exceptions thrown from `noexcept` functions called by special functions.

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/common/src/codingstandards/cpp/exceptions/ExceptionFlow.qll [hunk 1]
```diff
@@ -312,7 +313,21 @@ class ReThrowExprThrowingExpr extends ReThrowExpr, ThrowingExpr {
...
 /** An expression which calls a function which may throw an exception. */
 class FunctionCallThrowingExpr extends FunctionCall, ThrowingExpr {
-  override ExceptionType getAnExceptionType() { result = getAFunctionThrownType(getTarget(), _) }
+  override ExceptionType getAnExceptionType() {
+    exists(Function target |
+      target = getTarget() and
+      result = getAFunctionThrownType(target, _) and
+      // [expect.spec] states that throwing an exception type that is prohibited
+      // by the specification will result in the program terminating. We therefore
+      // do not propagate such exceptions to the call sites for the function.
```
Comment (rvermeulen) [inline cpp/common/src/codingstandards/cpp/exceptions/ExceptionFlow.qll:322]:
```suggestion
      // [expect.spec] states that throwing an exception type that is prohibited
      // by the specification will result in the program terminating, unless
      // a custom `unexpected_handler` is registered that throws an exception type
      // which is compatible with the specification.
      // We assume a default `std::unexpected` handler and therefore
      // do not propagate such exceptions to the call sites for the function.
      //
```

File: cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp [hunk 2]
```diff
@@ -1,9 +1,34 @@
...
+}
+
+void test_indirect_throw_5() noexcept {
+  noexceptIndirectThrowA(); // COMPLIANT - noexceptIndirectThrowA would call
+                            // std::terminate() if ExceptionA is thrown
+}
+
+void test_indirect_throw_6() noexcept {
+  indirectThrowA(); // NON_COMPLIANT
 }
```
Comment (rvermeulen) [inline cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp:34]:
```suggestion
}

void test_indirect_throw_6() noexcept {
    dynamicIndirectThrowA(); // NON_COMPLIANT[FALSE_NEGATIVE] - dynamicIndirectThrowA indirectly throws std::bad_exception
}
```

File: cpp/autosar/test/rules/A15-4-2/test.cpp [hunk 3]
```diff
@@ -21,4 +23,13 @@ void test_indirect_throw_3() noexcept(false) {
 
 void test_indirect_throw_4() {
   throwA(); // COMPLIANT - function marked as noexcept(false)
+}
+
+void test_indirect_throw_5() noexcept {
+  noexceptIndirectThrowA(); // COMPLIANT - noexceptIndirectThrowA would call
```
Comment (rvermeulen) [inline cpp/autosar/test/rules/A15-4-2/test.cpp:29]: A similar FN exists as for EXP55-CPP when a `std::bad_exception` can still be thrown.

**Review by rvermeulen — CHANGES_REQUESTED** [review]
Added a FN test case and suggested a comment change to capture an assumption. Not sure if we want to over approximate and rethrow `std::bad_exception` if that is part of the dynamic exception specification of a function that throws.

File: cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp (hunk 2 above)
Comment (lcartey) [reply cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp:34]: Ah, this is actually not a false negative. `dynamicIndirectThrowA` will not throw a `std::bad_exception` unless:

  * [`std::set_unexpected`](http://en.cppreference.com/w/cpp/error/set_unexpected) has been called with a custom unexpected error handler (note: this handler functionality was deprecated in C++11 and removed in C++17).
  * The custom handler re-throws an exception, and the exception it rethrows is not compatible with the dynamic exception specification.

This is shown in this example here: https://en.cppreference.com/w/cpp/error/bad_exception. Given this relates to deprecated functionality in our supported C++ standards (C++14 onwards), I don't think we need to implement this edge case or add a test for it. Let me know if you disagree.

Note: the [ERR55-CPP](https://wiki.sei.cmu.edu/confluence/display/cplusplus/ERR55-CPP.+Honor+exception+specifications) CERT rule has an incorrect statement about this behaviour:

> If a function declared with a dynamic-exception-specification throws an exception of a type that would not match the exception-specification, the function std::unexpected() is called **The behavior of this function can be overridden but, by default, causes an exception of std::bad_exception to be thrown**

That is not true - the default behaviour is for it to call `std::terminate`. You have to set the `unexpected` handler to trigger any of this behaviour.

File: cpp/common/src/codingstandards/cpp/exceptions/ExceptionFlow.qll (hunk 1 above)
Comment (lcartey) [reply cpp/common/src/codingstandards/cpp/exceptions/ExceptionFlow.qll:322]: Added, although I reworded it slightly for clarity.

File: cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp [hunk 4]
```diff
@@ -1,9 +1,34 @@
...
-  throw "test";
+void test_throw() noexcept(true) {
+  throw ExceptionA(); // NON_COMPLIANT - function marked as noexcept(true)
 }
 
-void test_noexcept_true() noexcept(true) { // NON_COMPLIANT
-  throw "test";
+void throwA() { throw ExceptionA(); }
+void indirectThrowA() { throwA(); }
+void noexceptIndirectThrowA() noexcept { throwA(); } // NON_COMPLIANT
```
[the remark this answers is not in the document]
Comment (lcartey) [reply cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp:9]: See below for rationale on not adding this case.

File: cpp/autosar/test/rules/A15-4-2/test.cpp (hunk 3 above)
Comment (lcartey) [reply cpp/autosar/test/rules/A15-4-2/test.cpp:29]: See below for rationale on not adding this case.

**Comment by lcartey** [comment]
Thanks for the review! As `std::bad_exception` is never thrown with the default handler (only via custom handlers), and the handler mechanism is deprecated in C++14, I have opted not to add the additional test cases - let me know if you disagree with that. I have clarified the comment in `ExceptionFlow.qll` to explain how it works and why we exclude it.

File: cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp (hunk 2 above)
Comment (rvermeulen) [reply cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp:34]: Agreed, it is only a FP in the case of a custom unexpected handler that rethrows an unspecified exception type. Good catch on the mistake in CERT!

**Review by rvermeulen — APPROVED** [review]
Given that the corner case is no longer possible in new standards it is time to :shipit: without.
